src/components/feedback/Modal.tsx

```diff
@@ -0,0 +1,136 @@
+import classnames from 'classnames';
+import type { RefObject } from 'preact';
+import { useEffect, useLayoutEffect } from 'preact/hooks';
+
+import { useElementShouldClose } from '../../hooks/use-element-should-close';
+import { useSyncedRef } from '../../hooks/use-synced-ref';
+import { useUniqueId } from '../../hooks/use-unique-id';
+
+import type { PresentationalProps } from '../../types';
+import { downcastRef } from '../../util/typing';
+
+import Overlay from '../layout/Overlay';
+import Panel from '../layout/Panel';
+import type { PanelProps } from '../layout/Panel';
+
+type ComponentProps = {
+  /**
+   * Modal _should_ be provided with a close handler. We have a few edge use
+   * cases, however, in which we need to render a "non-closeable" modal.
+   */
+  onClose?: () => void;
+  width?: 'sm' | 'md' | 'lg' | 'custom';
+  /**
+   * Element that should take focus when the Dialog is first rendered. When not
+   * provided, the dialog's outer element will take focus. Setting this prop to
+   * `null` will opt out of focus routing.
+   */
+  initialFocus?: RefObject<HTMLElement | null> | null;
+};
+
+export type ModalProps = PresentationalProps & ComponentProps & PanelProps;
+
+const noop = () => {};
+
+/**
+ * Show a modal
+ */
+const ModalNext = function Modal({
+  children,
+  width = 'md',
+
+  classes,
+  elementRef,
+  initialFocus,
+
+  // Forwarded to Panel
+  buttons,
+  icon,
+  onClose,
+  title,
+
+  ...htmlAttributes
+}: ModalProps) {
+  const modalRef = useSyncedRef(elementRef);
+  const closeHandler = onClose ?? noop;
+  useElementShouldClose(modalRef, true, closeHandler);
+
+  const dialogDescriptionId = useUniqueId('dialog-description');
+
+  useEffect(() => {
+    // Route focus unless `initialFocus` is set to `null`
+    if (initialFocus !== null) {
+      const focusEl = initialFocus?.current as HTMLInputElement;
```

Comment:
Hmm... the assumption that `focusEl` is an `HTMLInputElement` might not always be valid. For example it might be a textarea or button. I presume the reason we are doing this is that not all HTML elements have a `disabled` prop. We could make this clearer with something like:

```js
      const focusEl = initialFocus?.current as HTMLElement & { disabled?: boolean };
      if (focusEl && !focusEl.disabled) {
```

Reply:
> The combination of single line separator that doesn't cover the full width of the content, and top-shadow for scrolled content, which does, can look a bit odd:
> 
> <img alt="Top shadow and underline" width="662" src="https://user-images.githubusercontent.com/2458/216586796-5576d185-a9c2-46df-8dbb-4bd782ffee1a.png">

I tend to agree. However, this styling is actually owned by `Panel` so I want to hit this in a follow-up. Making a note.